### PR TITLE
Airway/Breathing/Circulation - Add ETCo2 Monitor

### PR DESCRIPTION
**Pull request by apo-tle**
**When merged this pull request will:**
- Add ETCo2 level to Vitals Monitor. ETCo2 level can be used to diagnose respiratory events such as airway blockages and PTX deterioration, as well as cardiac arrest
- Add Respiratory Rate to Vitals Monitor. Respiratory Rate can be used in conjunction to ETCo2 level to diagnose problems with a patient's airway.
- ETCo2 level and Respiratory Rate have no effect on any stats behind resuscitating a patient, instead serving as indicators to allow doctors to spot issues more easily.
- Add Nasal Cannula. Nasal Cannula will be required to be attached to a patient's head in order for ETCo2 and Respiratory Rate readouts to be visible on the Vitals Monitor. You will not be able to use the Nasal Cannula to provide a patient with oxygen.
### To Do
- [x] Implement Respiratory Rate
- [x] Implement Nasal Cannula
- [x] Remove ETCo2/RR readouts when Nasal Cannula is not connected
- [x] Add setting to allow ETCo2/RR readouts to be displayed
- [x] Remove debugging code and tidy up

### IMPORTANT

- [x] [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- [x] Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * [player] call kat_breathing_fnc_getETCo2;
+ *
+ * Public: No
+ */
+
+ params ["_patient"];
+
+if !(alive _patient) exitWith {0};
+
+_pr = _patient getVariable [QACEGVAR(medical,heartRate), 0];
```
Comment (MissHeda) [inline addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
private _pr = _patient getVariable [QACEGVAR(medical,heartRate), 80];
```

Comment (MissHeda) [reply addons/breathing/functions/fnc_getETCo2.sqf]: Also every where else in this file.

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+_bloodVolume = (_patient getVariable [QACEGVAR(medical,bloodVolume), 6.0]);
```
Comment (MissHeda) [inline addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
_bloodVolume = _patient getVariable [QACEGVAR(medical,bloodVolume), 6.0];
```

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+_lostBlood = 6.0 - _bloodVolume; // amount of blood missing from the body
+
+_ptxTarget = _patient getVariable [QGVAR(pneumothorax), 0]; // more deteriorated closer to 4, at 4 likely to become advanced
+_hasHtx = _patient getVariable [QGVAR(hemopneumothorax), false];
+_hasTptx = _patient getVariable [QGVAR(tensopneumothorax), false];
+
+_airwayObstructed = (_patient getVariable [QEGVAR(airway,obstruction), false]);
```
Comment (MissHeda) [inline addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
_airwayObstructed = _patient getVariable [QEGVAR(airway,obstruction), false];
```

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+_airwayOccluded = (_patient getVariable [QEGVAR(airway,occluded), false]);
```
Comment (MissHeda) [inline addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
_airwayOccluded = _patient getVariable [QEGVAR(airway,occluded), false];
```

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+//cardiac arrest
+if (_pr == 0) exitWith {0};
```
Comment (MissHeda) [inline addons/breathing/functions/fnc_getETCo2.sqf]: If this is the only time you use _pr then just use the code you used to set this variable.

Comment (MissHeda) [reply addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
if ((_patient getVariable [QACEGVAR(medical,heartRate), 80]) == 0) exitWith {0};
```

**Review by MissHeda — CHANGES_REQUESTED** [review]
Did not check everything in detail yet, just some fast sweep over.
I will add this to ongoing, let me or a dev know when you are done for another check.

File: addons/breathing/XEH_PREP.hpp [hunk 6]
```diff
@@ -21,4 +21,5 @@ PREP(treatmentAdvanced_pulseoximeterLocal);
 PREP(treatmentAdvanced_removePulseoximeter);
 PREP(treatmentAdvanced_tensionpneumothorax);
 PREP(treatmentAdvanced_tensionpneumothoraxLocal);
-PREP(useBVM);
\ No newline at end of file
+PREP(useBVM);
+PREP(getETCo2);
```
Comment (MiszczuZPolski) [inline addons/breathing/XEH_PREP.hpp]: fnc in PREP goes alphabetically

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 7]
```diff
@@ -0,0 +1,61 @@
+#include "..\script_component.hpp"
+/*
+ * Author: apo_tle
+ * Calculate patient's ETCo2 level based on status of their vitals. ETCo2 is not its own vital as it does not need to be,
```
Comment (MiszczuZPolski) [inline addons/breathing/functions/fnc_getETCo2.sqf]: delete unnecessary description, "Calculate patient's ETCo2 level based on status of their vitals." is enough

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ * Return Value:
+ * ETCo2 level <INT>
+ *
+ * Example:
+ * [player] call kat_breathing_fnc_getETCo2;
+ *
+ * Public: No
+ */
+
+ params ["_patient"];
```
Comment (MiszczuZPolski) [inline addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
params ["_patient"];
```

File: addons/breathing/functions/fnc_getETCo2.sqf (hunk 1 above)
Comment (MiszczuZPolski) [reply addons/breathing/functions/fnc_getETCo2.sqf]: you can use macros too, i think it should work without issues
```suggestion
private _pr = GET_HEART_RATE(_patient);
```

File: addons/breathing/functions/fnc_getETCo2.sqf (hunk 2 above)
Comment (MiszczuZPolski) [reply addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
private _bloodVolume = GET_BLOOD_VOLUME(_patient);
```

File: addons/breathing/functions/fnc_getETCo2.sqf (hunk 5 above)
Comment (MiszczuZPolski) [reply addons/breathing/functions/fnc_getETCo2.sqf]: If you are using that to check if person is in cardiac arrest then i suggest
```suggestion
if (_IN_CRDC_ARRST(_patient)) exitWith {0};
```

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+_hasTptx = _patient getVariable [QGVAR(tensopneumothorax), false];
+
+_airwayObstructed = (_patient getVariable [QEGVAR(airway,obstruction), false]);
+_airwayOccluded = (_patient getVariable [QEGVAR(airway,occluded), false]);
+
+//cardiac arrest
+if (_pr == 0) exitWith {0};
+
+//cpr being performed
+if !(_patient getVariable [QACEGVAR(medical,CPR_provider), objNull] isEqualTo objNull) exitWith {
```
Comment (MiszczuZPolski) [inline addons/breathing/functions/fnc_getETCo2.sqf]:
```suggestion
if (alive (_patient getVariable [QACEGVAR(medical,CPR_provider), objNull])) exitWith {
```

**Review by MiszczuZPolski — CHANGES_REQUESTED** [review]
I would use switch case in getETCo2 instead checking each var with if

File: addons/breathing/functions/fnc_getETCo2.sqf (hunk 9 above)
Comment (apo-tle) [reply addons/breathing/functions/fnc_getETCo2.sqf]: No need as we already check the patient is alive earlier in the script

Comment (MiszczuZPolski) [reply addons/breathing/functions/fnc_getETCo2.sqf]: I basically used what ACE does to check if there is someone providing CPR in CPRLocal

**Comment by apo-tle** [comment]
### To Do for Nasal Cannula

- [x] Add Nasal Cannula label to head quick view
- [x] Add Nasal Cannula to log when inserted
- [x] Change VM so ETCo2 level only appears with nasal cannula attached
- [x] UI element for Nasal Cannula (need support for this)

**Comment by Ghostworrior** [comment]
Etco2 should display with either nasal cannula or KingLT in place (if I or someone else ever gets to do the airway rework) other items would be added as well.

Etco2 under good CPR should be ~15 rather than the 10 it is currently. With high blood loss, etco2 should drop even under CPR to 5-10.
If CPR is performed on a patient with circulation it should still show 35-45 as a value, but RR should be fixed at 10 ventilations while doing CPR or using a BVM.

Small idea on the side increase etco2 and decrease RR if lorazepam, fentanyl or morphine is used. Not 100% realistic, but easiest way to implement/show side effects.

Also change etco2 and RR in correlation to pH if it's in use, the lower pH value the higher RR and etco2.

Just my two cents after looking over the code, all in all I like the system, good work.
Will gladly test it in game once It's either ready or I figure out how to do Git stuff again...

**Comment by apo-tle** [comment]
> Etco2 under good CPR should be ~15 rather than the 10 it is currently. With high blood loss, etco2 should drop even under CPR to 5-10. If CPR is performed on a patient with circulation it should still show 35-45 as a value, but RR should be fixed at 10 ventilations while doing CPR or using a BVM.

Currently etco2 under CPR changes based on how deteriorated a patient's cardiac arrest is, starting at 10 and increasing to 16 as compressions move from asystole -> VT. My idea behind this was to simulate etco2 continually increasing as good cpr is performed. I'll adjust values to decrease etco2 further for blood loss.

**Comment by apo-tle** [comment]
ETCo2 Monitor is all done and ready for a final review.

File: addons/breathing/functions/fnc_getETCo2.sqf [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+private _phDiff = 1500 - _pH;
+
+private _ptxTarget = _patient getVariable [QGVAR(pneumothorax), 0]; // more deteriorated closer to 4, at 4 likely to become advanced
+private _hasHtx = _patient getVariable [QGVAR(hemopneumothorax), false];
+private _hasTptx = _patient getVariable [QGVAR(tensopneumothorax), false];
+
+private _airwayObstructed = _patient getVariable [QEGVAR(airway,obstruction), false] && !(_patient getVariable [QEGVAR(airway,overstretch), false]);
+private _airwayOccluded = _patient getVariable [QEGVAR(airway,occluded), false];
+
+private _phOffset = [0,(floor (_phDiff / 250))] select (EGVAR(pharma,kidneyAction)); // increase etco2 by 1 per 250 points of pH lost
```
Comment (BrettMayson) [inline addons/breathing/functions/fnc_getETCo2.sqf]: Will break if pharma is not loaded

**Review by BrettMayson — CHANGES_REQUESTED** [review]

File: addons/breathing/functions/fnc_getETCo2.sqf (hunk 10 above)
Comment (apo-tle) [reply addons/breathing/functions/fnc_getETCo2.sqf]: Would an if statement checking for pharma being enabled fix this or would it still break?

**Comment by mazinskihenry** [comment]
LGTM

**Pull request merged**